pkg/apis/apis/v1alpha1/types_apiresourceschema.go

```diff
@@ -163,3 +163,104 @@ func (v *APIResourceVersion) SetSchema(schema *apiextensionsv1.JSONSchemaProps)
 	v.Schema.Raw = raw
 	return nil
 }
+
+const (
+	// VersionPreservationAnnotationKeyPrefix is the prefix for the annotation key used to preserve fields from an API
+	// version that would otherwise be lost during round-tripping to a different API version. An example key and value
+	// might look like this: preserve.conversion.apis.kcp.io/v2: {"spec.someNewField": "someValue"}.
+	VersionPreservationAnnotationKeyPrefix = "preserve.conversion.apis.kcp.io/"
+)
+
+// +crd
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Cluster,categories=kcp
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+
+// APIConversion contains rules to convert between different API versions in an APIResourceSchema. The name must match
+// the name of the APIResourceSchema for the conversions to take effect.
+type APIConversion struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	// Spec holds the desired state.
+	Spec APIConversionSpec `json:"spec"`
+}
+
+// APIConversionSpec contains rules to convert between different API versions in an APIResourceSchema.
+type APIConversionSpec struct {
+	// +required
+	// +listType=map
+	// +listMapKey=from
+	// +listMapKey=to
+
+	// conversions specify rules to convert between different API versions in an APIResourceSchema.
+	Conversions []APIVersionConversion `json:"conversions"`
+}
+
+// APIVersionConversion contains rules to convert between two specific API versions in an
+// APIResourceSchema. Additionally, to avoid data loss when round-tripping from a version that
+// contains a new field to one that doesn't and back again, you can specify a list of fields to
+// preserve (these are stored in annotations).
+type APIVersionConversion struct {
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Pattern=^v[1-9][0-9]*([a-z]+[1-9][0-9]*)?$
+
+	// from is the source version.
+	From string `json:"from"`
```

Comment:
```suggestion
        // from is the source version.
        //
	// +required
	// +kubebuilder:validation:MinLength=1
	// +kubebuilder:validation:Pattern=^v[1-9][0-9]*([a-z]+[1-9][0-9]*)?$
	From string `json:"from"`
```

Is there a reason why we don't prefer the pattern I've seen above? Having everything in the same block makes it easier (at least for me, but others might disagree) to parse which rules are associated with which field, and keeps everything together.

Reply:
Godoc generation. If you keep it all together, the directives are included in the doc